core/peerset/src/peersstate.rs

```diff
@@ -0,0 +1,586 @@
+			.find(|(_, &mut Node { connection_state, reserved, .. })| {
+				reserved && !connection_state.is_connected()
+			})
+			.map(|(peer_id, node)| (peer_id.clone(), node));
+
+		if let Some((peer_id, state)) = outcome {
+			Some(NotConnectedPeer {
+				state,
+				peer_id: Cow::Owned(peer_id),
+				num_in: &mut self.num_in,
+				num_out: &mut self.num_out,
+				max_in: self.max_in,
+				max_out: self.max_out,
+			})
+		} else {
+			None
+		}
+	}
+
+	/// Returns the peer with the highest reputation and that we are not connected to.
+	///
+	/// If multiple nodes have the same reputation, which one is returned is unspecified.
+	pub fn highest_not_connected_peer(&mut self) -> Option<NotConnectedPeer> {
+		let outcome = self.nodes
+			.iter_mut()
+			.filter(|(_, Node { connection_state, .. })| !connection_state.is_connected())
+			.fold(None::<(&PeerId, &mut Node)>, |mut cur_node, to_try| {
+				if let Some(cur_node) = cur_node.take() {
+					if cur_node.1.reputation >= to_try.1.reputation {
+						return Some(cur_node);
+					}
+				}
+				Some(to_try)
+			})
+			.map(|(peer_id, state)| (peer_id.clone(), state));
+
+		if let Some((peer_id, state)) = outcome {
+			Some(NotConnectedPeer {
+				state,
+				peer_id: Cow::Owned(peer_id),
+				num_in: &mut self.num_in,
+				num_out: &mut self.num_out,
+				max_in: self.max_in,
+				max_out: self.max_out,
+			})
+		} else {
+			None
+		}
+	}
+}
+
+/// Grants access to the state of a peer in the `PeersState`.
+pub enum Peer<'a> {
+	/// We are connected to this node.
+	Connected(ConnectedPeer<'a>),
+	/// We are not connected to this node.
+	NotConnected(NotConnectedPeer<'a>),
+	/// We have never heard of this node.
+	Unknown(UnknownPeer<'a>),
+}
+
+impl<'a> Peer<'a> {
+	/// If we are the `Connected` variant, returns the inner `ConnectedPeer`. Returns `None`
+	/// otherwise.
+	pub fn into_connected(self) -> Option<ConnectedPeer<'a>> {
+		match self {
+			Peer::Connected(peer) => Some(peer),
+			Peer::NotConnected(_) => None,
+			Peer::Unknown(_) => None,
+		}
+	}
+
+	/// If we are the `Unknown` variant, returns the inner `ConnectedPeer`. Returns `None`
+	/// otherwise.
+	#[cfg(test)]	// Feel free to remove this if this function is needed outside of tests
+	pub fn into_not_connected(self) -> Option<NotConnectedPeer<'a>> {
+		match self {
+			Peer::Connected(_) => None,
+			Peer::NotConnected(peer) => Some(peer),
+			Peer::Unknown(_) => None,
+		}
+	}
+
+	/// If we are the `Unknown` variant, returns the inner `ConnectedPeer`. Returns `None`
+	/// otherwise.
+	#[cfg(test)]	// Feel free to remove this if this function is needed outside of tests
+	pub fn into_unknown(self) -> Option<UnknownPeer<'a>> {
+		match self {
+			Peer::Connected(_) => None,
+			Peer::NotConnected(_) => None,
+			Peer::Unknown(peer) => Some(peer),
+		}
+	}
+}
+
+/// A peer that is connected to us.
+pub struct ConnectedPeer<'a> {
+	state: &'a mut Node,
+	peer_id: Cow<'a, PeerId>,
+	num_in: &'a mut u32,
+	num_out: &'a mut u32,
+	max_in: u32,
```

Comment:
What are these? Do they need to be stored for each peer?

Reply:
I added them in the latest commit to address Gav's review about the panic proof not being strong enough. We now store a `&mut Node` instead of the `&mut Peerstate`, but that means we copy over some fields of the `Peerstate`.